craft_application/services/proxy.py

```diff
@@ -0,0 +1,165 @@
+    """True if the proxy service has been configured."""
+
+    @final
+    def configure(self, proxy_cert: pathlib.Path, http_proxy: str) -> None:
+        """Configure the proxy service.
+
+        :param proxy_cert: The path to the proxy certificate to install in the instance.
+        :param http_proxy: The proxy url to set in the instance.
+        """
+        self.__proxy_cert = proxy_cert
+        self.__http_proxy = http_proxy
+        self.__is_configured = True
+
+    @final
+    def configure_instance(self, instance: craft_providers.Executor) -> dict[str, str]:
+        """Configure a build instance to connect to a proxy.
+
+        :param instance: The instance to configure.
+
+        :returns: A dict of environment variables to set in the instance.
+        """
+        if not self.__is_configured:
+            emit.debug(
+                "Skipping proxy configuration because the proxy service isn't configured."
+            )
+            return {}
+
+        emit.progress("Configuring proxy in instance.")
+
+        self._install_certificate(instance)
+        self._configure_pip(instance)
+        self._configure_snapd(instance)
+        self._configure_apt(instance)
+
+        return self._env
+
+    @property
+    def _env(self) -> dict[str, str]:
+        """Environment variables to use for the proxy."""
+        return {
+            "http_proxy": self.__http_proxy,
+            "https_proxy": self.__http_proxy,
+            # This makes the requests lib take our cert into account.
+            "REQUESTS_CA_BUNDLE": str(_PROXY_CERT_INSTANCE_PATH),
+            # Same, but for cargo.
+            "CARGO_HTTP_CAINFO": str(_PROXY_CERT_INSTANCE_PATH),
+            # Have go download directly from repositories
+            "GOPROXY": "direct",
+        }
+
+    def _configure_pip(self, instance: craft_providers.Executor) -> None:
+        emit.progress("Configuring pip")
+
+        self._execute_run(instance, ["mkdir", "-p", "/root/.pip"])
+        pip_config = b"[global]\ncert=/usr/local/share/ca-certificates/local-ca.crt"
+        instance.push_file_io(
+            destination=pathlib.Path("/root/.pip/pip.conf"),
+            content=io.BytesIO(pip_config),
+            file_mode="0644",
+        )
+
+    def _configure_snapd(self, instance: craft_providers.Executor) -> None:
+        """Configure snapd to use the proxy and see our certificate.
+
+        Note: This must be called after _install_certificate(), to ensure that
+        when the snapd restart happens the new cert is there.
+        """
+        emit.progress("Configuring snapd")
+        self._execute_run(instance, ["systemctl", "restart", "snapd"])
+        for config in ("proxy.http", "proxy.https"):
+            self._execute_run(
+                instance, ["snap", "set", "system", f"{config}={self.__http_proxy}"]
+            )
+
+    def _configure_apt(self, instance: craft_providers.Executor) -> None:
+        emit.progress("Configuring Apt")
```

Comment:
```suggestion
        try:
            instance.execute_run(["test", "-d", "/etc/apt"], check=True)
        except subprocess.CalledProcessError:
            emit.debug("Apt not available in instance. Skipping configuration.")
            return
        emit.progress("Configuring Apt")
```

Reply:
Coming in a separate PR in a few moments :)

Reply:
https://github.com/canonical/craft-application/pull/824